src/compiler/checker.ts

```diff
@@ -8915,13 +8949,31 @@ namespace ts {
                 type.resolvedIndexType || (type.resolvedIndexType = createIndexType(type, /*stringsOnly*/ false));
         }
 
+        function guardedBigInt(value: string | number, node?: Node): number {
+            if (typeof BigInt === "undefined") { // error and fall back to number parsing
+                error(node, Diagnostics.Cannot_evaluate_bigint_literal_because_the_compiler_is_running_in_an_environment_without_bigint_support);
+                return +value;
+            }
+            return BigInt(value);

```

Comment:
We talked about this internally when working on some other stuff; but I think similarly to our reluctance to rely on host features for any potential regex types, it's much too soon for us to rely on a host's `BigInt` constructor to be present to handle `BigInt` types in the compiler. Unlike the runtime, we don't need to perform arithmetic or much of anything on these types - the only operation we (potentially) need is equivalence. Our scanner already does octal/hex -> value conversions; we should just be able to extend that to handle integer overflow/use a psudobigint representation for large numbers and do our equivalence checking internally with a custom partial bigint.

Reply:
If `bigint` literals remain valid types, you then I think you need some sort of arbitrary-precision integer library in order to compare ones in different bases, as literals can overflow the 53-bit integer available with the `number` type. It would probably be safe to just compute the lower 32 bits or something, but that isn't technically correct. Also, `evaluate` attempts to evaluate arithmetic operations on numeric literals, so do you think that should just be disabled for `bigint` literals?

Reply:
Perhaps restrict `bigint` literals to the range which can be represented absolutely by the `number` type?

Reply:
@calebsander You don't need a platform primitive for that, though. We read everything in as strings; we can read a bigint in as a string, then convert its value into a buffer of bytes (or another mroe efficient representation, since bigints tend to be sparse) _ourselves_, then compare those buffers, rather than relying on the host.

Reply:
> Also, evaluate attempts to evaluate arithmetic operations on numeric literals, so do you think that should just be disabled for bigint literals?

Yeah, I'd just leave that off - that's just for enums, and const bigint enums aren't a pressing concern.

Reply:
I'd also say that we want to be lazy here if it all possible - if a user only ever writes `0n`, then string comparisons are all we ever need - it's only when both `0x0n` and `0o0n` see use, for example, that we need to actually start parsing and comparing literal values. Necessary, for sure, but we _really_ want to avoid doing it for numbers outside the typical double range. Actually, since it seems like bigints don't support scientific notation (probably to avoid the whole negative exponent issue), there's only ever exactly 4 possible strings representing a given number - the base 2 string, the base 8 string, the base 16 string, and the base 10 string. I think if we normalize a bigint "value" internally as "the base 64 string representation of the value" (base64 simply because it's a pretty common way to encode buffers into strings - maybe another base with compression is better for memory; without profiling on a base that makes heavy use of large literals I wouldn't know), then string equality will suffice; we just need to bulk up the scanner to be able to perform that conversion.